### PR TITLE
Add support for multiple conditions to target a common field

### PR DESCRIPTION
**Pull request by brunobord**
## Review

This PR uses the same test suite as defined in https://github.com/peopledoc/django-formidable/pull/299

A few notes:

At the moment, ``Conditions`` are only used for defining conditional display of fields. But if we inspect its implementation, we'll find that it's being built around a bigger scope, that is to say that they're built to be as "generic" as possible. Virtually, ``Condtions`` could lead to a variety of actions.

In the meantime, the ``display_iff`` action defined in the present code had a definitive impact on the form: having only *one* "hide this field" action would **wipe it out** from the form fields and ``cleaned_data`` and ``errors``... My opinion on this, regardless of the evolution we're trying to implement here, is that the ``Condition`` object was way too powerful.

The required evolution was conflicting with these design decisions.

The first "naive" implementation would be to redefine ``Condition`` objects as "Display Conditions" only, and the only available action would be to flag a field to be displayed or not.
On top of this first step, we'll open the path to other "extra actions". My opinion is that, if one day a ``Condition`` defines a new action (the example provided by @charlineb was *"flag a field as required"*), it'll also make it **mandatory** that this field **was displayed** in the first place.
So, IMHO, I think it's safe to make sure that each targeted field will be displayed because at least one ``Condition`` says so.
To sum this up: if one day there's a ``"SetToRequiredCondition"`` class, it'll have to use the same ``keep_fields`` method defined here, and eventually apply its action(s) on the fields that are still there.

But at the moment, we only have one kind of Condition/Action. In order to avoid the [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) trap, I won't define extra actions or the "apply" mechanism yet.

* [ ] Tests
* [ ] Docs/comments
* [ ] `CHANGELOG.rst` Updated
* [ ] Delete your branch

### CODE REVIEW

File: docs/source/forms.rst [hunk 1]
```diff
@@ -189,6 +189,17 @@ Conditions
...
+
+    As of ``1.4.0``, it is allowed to have several conditional display rules that target a common field. In case of "conflict" between these rules, priority goes to the *display*, rather than the *hide* action.
+
+    e.g.:
+
+    * Rule 1 says: "if checkbox-1 is checked, then display field X"
+    * Rule 2 says: "if checkbox-2 is checked, then display field X and Y"
+
+    if only checkbox-1 is checked, field X will be displayed, even if checkbox-2 is unchecked, and vice-versa. If both are checked, fields X and Y won't be displayed. If none is checked, fields X and Y will be hidden.
+
```
Comment (alexdashkov) [inline docs/source/forms.rst]: It seems that there is a typo here
`fields X and Y won't be displayed` -> `fields X and Y will be displayed`

File: formidable/forms/__init__.py [hunk 2]
```diff
@@ -48,10 +48,45 @@ def __init__(self, *args, **kwargs):
...
+                condition_targets[field_id].append(keep_fields)
+        # Here, the catalog contains fields targeted by 1 or many conditions.
+
+        # If only one condition says "please display X", we'll keep X
+        # That's why we gather the conditions using "any"
+        condition_targets = {k: any(v) for k, v in condition_targets.items()}
+        # We'll only remove fields that are targeted by conditions **and**
+        # those conditions are false
+        to_remove = [k for k, v in condition_targets.items() if not v]
+        return to_remove
```
Comment (alexdashkov) [inline formidable/forms/__init__.py]: I would like to merge the lines 71 and 74. 
`to_remove = [k for k, v in condition_targets.items() if not any(v)]`
or just
`return [k for k, v in condition_targets.items() if not any(v)] `

File: demo/tests/test_end_point.py [hunk 3]
```diff
@@ -915,11 +915,9 @@ def test_create_form_conditions_invalid_dup(self):
         data['fields'] = copy.deepcopy(self.fields_with_validation)
         data['conditions'] = copy.deepcopy(self.valid_conditions_invalid_dup)
         serializer = FormidableSerializer(data=data)
-        self.assertFalse(serializer.is_valid())
-        self.assertEqual(
-            serializer.errors['non_field_errors'][0],
-            'Action display_iff in condition (my condition) is used many times for the same fields (input-date)' # noqa
-        )
+        # TODO decide if the validation should fail
```
Comment (brunobord) [inline demo/tests/test_end_point.py]: not sure if I want to keep this TODO here

File: formidable/forms/__init__.py [hunk 4]
```diff
@@ -48,10 +48,45 @@ def __init__(self, *args, **kwargs):
...
+
     def clean(self):
         cleaned_data = super(BaseDynamicForm, self).clean()
-        for condition in self._conditions:
-            cleaned_data = condition(self, cleaned_data)
+
+        removed_fields = self.get_removed_fields(cleaned_data)
```
Comment (alexdashkov) [inline formidable/forms/__init__.py]: And I think that we can use a `generator` here instead of a `list`

File: demo/tests/test_end_point.py (hunk 3 above)
Comment (alexdashkov) [reply demo/tests/test_end_point.py]: BTW, why an action name is `display_iff` and not just a `display_if`?

Comment (syldlb) [reply demo/tests/test_end_point.py]: I was unsure about this one: we have the exact same condition two times. Should it be allowed? Either for displaying or making mandatory it does't impact the form so I would say yes.

Then this test should be renamed with "valid" instead of "invalid", and the "valid_conditions_invalid_dup" in "valid_conditions_with_dup".

Comment (brunobord) [reply demo/tests/test_end_point.py]: > BTW, why an action name is display_iff and not just a display_if?

I don't know. I think we may ask @mgu

Comment (brunobord) [reply demo/tests/test_end_point.py]: ok, so I'll let this "TODO" and we'll work on it in an other task. @syldlb could you open a github issue in this repository, please?

Comment (mgu) [reply demo/tests/test_end_point.py]: https://en.wikipedia.org/wiki/iff

usual mathematical abbreviation

File: formidable/forms/__init__.py (hunk 2 above)
Comment (brunobord) [reply formidable/forms/__init__.py]: we could. and at the same time, I think that *readability counts*. this comprehension list is not hard to read now, but it could be handy to see how it is composed using two steps, reading comments and being able to inspect the values at runtime, when we'll have to debug... within a few months time.

File: formidable/forms/__init__.py (hunk 4 above)
Comment (brunobord) [reply formidable/forms/__init__.py]: Fixed.

File: docs/source/forms.rst (hunk 1 above)
Comment (brunobord) [reply docs/source/forms.rst]: oooops, fixed.

File: demo/tests/test_end_point.py (hunk 3 above)
Comment (mgu) [reply demo/tests/test_end_point.py]: but as you are changing the way it works, it makes no sense to keep this name…

Comment (brunobord) [reply demo/tests/test_end_point.py]: I see... it'd be cool to make a data migration using the JSON migrations, but not here yet ; changing the name of the only action available is not worth it IMO.

File: formidable/forms/__init__.py [hunk 5]
```diff
@@ -48,10 +48,44 @@ def __init__(self, *args, **kwargs):
...
+                    condition_targets[field_id] = []
+                condition_targets[field_id].append(keep_fields)
+        # Here, the catalog contains fields targeted by 1 or many conditions.
+
+        # If only one condition says "please display X", we'll keep X
+        # That's why we gather the conditions using "any"
+        condition_targets = {k: any(v) for k, v in condition_targets.items()}
+        # We'll only remove fields that are targeted by conditions **and**
+        # those conditions are false
+        return (k for k, v in condition_targets.items() if not v)
```
Comment (alexdashkov) [inline formidable/forms/__init__.py]: So you don't want to remove intermediate variable? And write just a 
`(k for k, v in condition_targets.items() if not any(v)`

Comment (brunobord) [reply formidable/forms/__init__.py]: IMHO the intermediate variable is the way to make sure that you can understand what it does. It's very easy to overlook a comprehension list and misread. ``if not v`` and ``if not any(v)`` are very close. Code is more often read than written.
If you have two filters/actions to apply to an iterable, I prefer having two lines of code, each one doing one thing and doing it clearly.
And one can inspect the intermediate variable at debugging time.

Comment (alexdashkov) [reply formidable/forms/__init__.py]: Ok, I see. Maybe it's just because I'm in love with this python comprehension list/dict feature and I don't like to add extra lines in this case :)

Comment (brunobord) [reply formidable/forms/__init__.py]: :)
